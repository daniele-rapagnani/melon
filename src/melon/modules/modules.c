#include "melon/modules/modules.h"
#include "melon/core/utils.h"
#include "melon/core/gc.h"
#include "melon/core/tstring.h"
#include "melon/core/compiler.h"
#include "melon/core/object.h"
#include "melon/core/closure.h"
#include "melon/core/program.h"
#include "melon/core/deserializer.h"

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <assert.h>

TRet melNewModule(VM* vm, const ModuleFunction* functions)
{
    if (functions == NULL)
    {
        return 1;
    }

    GCItem* libObj = melNewObject(vm);
    melM_vstackPushGCItem(&vm->stack, libObj);

    const ModuleFunction* curFunc = functions;

    melSaveAndPauseGC(vm, &vm->gc);

    while (curFunc->func != NULL)
    {
        GCItem* fObj = melNewFunction(vm);
        Function* f = melM_functionFromObj(fObj);

        f->localsSlots = curFunc->localSlots;
        f->args = curFunc->args;
        f->native = curFunc->func;
        f->method = curFunc->method;
        f->rest = curFunc->rest;

        if (curFunc->name != NULL)
        {
            f->name = melNewString(vm, curFunc->name, strlen(curFunc->name));
        }

        melAddFunctionToObjModule(
            vm, 
            libObj, f->name != NULL ? f->name : MELON_SYMBOLIC_KEYS[MELON_OBJSYM_CALL].pack.obj, 
            fObj
        );

        curFunc++;
    }

    melRestorePauseGC(vm, &vm->gc);

    return 0;
}

struct ErrorContext
{
    GCItem* callback;
    GCItem* file;
    VM* vm;
};

static void errorCb(
    void* ctx, 
    const char* message,
    const char* file,
    TSize line, 
    TSize column
)
{
    struct ErrorContext* errCtx = (struct ErrorContext*)ctx;
    assert(errCtx != NULL);
    assert(errCtx->callback != NULL);
    assert(errCtx->vm != NULL);

    melM_vstackPushGCItem(&errCtx->vm->stack, errCtx->callback);
    melM_vstackPushGCItem(&errCtx->vm->stack, melNewString(errCtx->vm, message, strlen(message)));

    if (file != NULL)
    {
        melM_vstackPushGCItem(&errCtx->vm->stack, melNewString(errCtx->vm, file, strlen(file)));
    }
    else
    {
        melM_vstackPushNull(&errCtx->vm->stack);
    }

    Value tmp;
    tmp.type = MELON_TYPE_INTEGER;

    tmp.pack.value.integer = line;
    melM_stackPush(&errCtx->vm->stack, &tmp);
    tmp.pack.value.integer = column;
    melM_stackPush(&errCtx->vm->stack, &tmp);

    melCallClosureSyncVM(errCtx->vm, 4, 0, 0);
}

GCItem* melCompileSourceModule(VM* vm, GCItem* path, GCItem* src, GCItem* errorCallback)
{
    String* strObj = melM_strFromObj(src);
    const char* strData = melM_strDataFromObj(src);
    const char* pathData = path != NULL ? melM_strDataFromObj(path) : NULL;

    Compiler c;

    if (melCreateCompilerFile(&c, vm, pathData, strData, strObj->len) != 0)
    {
        melM_fatal(vm, "Error while creating compiler.");
        return NULL;
    }

    c.module = 1;

    struct ErrorContext* ctx = NULL;

    if (errorCallback != NULL)
    {
        ctx = malloc(sizeof(struct ErrorContext));
        ctx->vm = vm;
        ctx->callback = errorCallback;

        // The compiler will provide its own file
        ctx->file = NULL;

        c.errorCtx = ctx;
        c.error = errorCb;

        c.lexer.errorCtx = ctx;
        c.lexer.error = errorCb;
    }

    if (melRunCompiler(&c) != 0)
    {
        free(ctx);
        return NULL;
    }

    free(ctx);

    return c.main.func;
}

static void errorDeserializerCb(
    void* ctx, 
    const char* message
)
{
    errorCb(ctx, message, NULL, 0, 0);
}

GCItem* melCompileBytecodeModule(VM* vm, GCItem* path, GCItem* code, GCItem* errorCallback)
{
    Deserializer ds;

    struct ErrorContext* ctx = NULL;

    if (errorCallback != NULL)
    {
        ctx = malloc(sizeof(struct ErrorContext));
        ctx->vm = vm;
        ctx->callback = errorCallback;
        ctx->file = path;

        ds.ctx = ctx;
        ds.error = errorDeserializerCb;
    }

    String* codeStr = melM_strFromObj(code);
    TByte* data = (TByte*)melM_strDataFromObj(code);
    GCItem* prog = NULL;

    const char* pathData = melM_strDataFromObj(path);

    if (melLoadProgramVM(vm, data, codeStr->len, &prog, pathData, &ds) != 0)
    {
        free(ctx);
        return NULL;
    }
    
    free(ctx);

    Program* p = melM_programFromObj(prog);
    return p->main;
}

static TBool isByteCode(VM* vm, GCItem* src)
{
    String* srcStr = melM_strFromObj(src);

    if (srcStr->len < MELON_VM_MAGIC_BYTES_COUNT)
    {
        return 0;
    }

    TByte* data = (TByte*)melM_strDataFromObj(src);

    return memcmp(data, &MAGIC_BYTES, MELON_VM_MAGIC_BYTES_COUNT) == 0;
}

TRet melNewModuleFromSource(VM* vm, GCItem* path, GCItem* src, GCItem* errorCallback)
{
    assert(src->type == MELON_TYPE_STRING);
    assert(path == NULL || path->type == MELON_TYPE_STRING);

    GCItem* mainFunc = NULL;

    if (isByteCode(vm, src))
    {
        mainFunc = melCompileBytecodeModule(vm, path, src, errorCallback);
    }
    else
    {
        mainFunc = melCompileSourceModule(vm, path, src, errorCallback);
    }

    if (mainFunc == NULL)
    {
        return 1;
    }

    // Prevent GCing
    melM_vstackPushGCItem(&vm->stack, mainFunc);

    Value progCl;
    progCl.type = MELON_TYPE_CLOSURE;
    progCl.pack.obj = melNewClosure(vm, mainFunc);

    melM_stackPop(&vm->stack);
    melM_stackPush(&vm->stack, &progCl);

    return 0;
}

TRet melRegisterModule(VM* vm, const char* name, ModuleInitFunc initFunc)
{
    assert(initFunc != NULL);

    initFunc(vm);

    StackEntry* obj = melM_stackOffset(&vm->stack, 0);
    assert(obj->type == MELON_TYPE_OBJECT);

#ifdef _DEBUG_GC
    printf("[Modules] Module registered '%s': %p\n", name, obj->pack.obj);
#endif

    GCItem* nameString = melNewString(vm, name, strlen(name));
    melM_vstackPushGCItem(&vm->stack, nameString); // Avoid GCing

    melAddItemToGlobalModule(vm, nameString, obj->pack.obj);

    melM_stackPop(&vm->stack); // Module name
    melM_stackPop(&vm->stack); // Module obj

    return 0;
}

TRet melRegisterModules(VM* vm, const Module* modules)
{
    const Module* curMod = modules;

    while (curMod->name != NULL)
    {
        if (melRegisterModule(vm, curMod->name, curMod->initFunc) != 0)
        {
            return 1;
        }

        curMod++;
    }

    return 0;
}

TRet melAddValueToObjModule(VM* vm, GCItem* obj, GCItem* name, Value* v)
{
    Value keyName;
    keyName.type = name->type;
    keyName.pack.obj = name;

    return melSetValueObject(vm, obj, &keyName, v);
}

TRet melAddItemToObjModule(VM* vm, GCItem* obj, GCItem* name, GCItem* i)
{
    Value val;
    val.type = i->type;
    val.pack.obj = i;

    return melAddValueToObjModule(vm, obj, name, &val);
}

TRet melAddFunctionToObjModule(VM* vm, GCItem* obj, GCItem* name, GCItem* fn)
{
    GCItem* closure = melNewClosure(vm, fn);
    return melAddItemToObjModule(vm, obj, name, closure);
}

TRet melAddValueToGlobalModule(VM* vm, GCItem* name, Value* v)
{
    return melAddValueToObjModule(vm, vm->global, name, v);
}

TRet melAddFunctionToGlobalModule(VM* vm, GCItem* name, GCItem* fn)
{
    return melAddFunctionToObjModule(vm, vm->global, name, fn);
}

TRet melAddItemToGlobalModule(VM* vm, GCItem* name, GCItem* item)
{
    return melAddItemToObjModule(vm, vm->global, name, item);
}

Value* melGetValueFromGlobalModule(VM* vm, Value* key)
{
    return melGetValueObject(vm, vm->global, key);
}

TRet melCreateKeyModule(VM* vm, const char* key, Value* value)
{
    value->type = MELON_TYPE_STRING;
    value->pack.obj = melNewString(vm, key, strlen(key));
    melAddRootValueGC(vm, &vm->gc, value);

    return 0;
}

Value* melGetTypeFromModule(VM* vm, Value* moduleName, Value* keyName, TType type)
{
    GCItem* module = vm->global;

    if (moduleName->type != MELON_TYPE_NULL)
    {
        Value* moduleVal = melGetValueFromGlobalModule(vm, moduleName);
        module = moduleVal->pack.obj;
    }
    
    if (module == NULL || module->type != MELON_TYPE_OBJECT)
    {
        return NULL;
    }

    Value* val = melGetValueObject(vm, module, keyName);

    if (val == NULL || val->type != type)
    {
        return NULL;
    }

    return val;
}

Value* melGetClosureFromModule(VM* vm, Value* moduleName, Value* funcName)
{
    return melGetTypeFromModule(vm, moduleName, funcName, MELON_TYPE_CLOSURE);
}

Value* melGetArgModule(VM* vm, const char* name, TSize slot, TType type, TBool optional)
{
    Value* arg = melGetLocalVM(vm, melGetTopCallFrameVM(vm), slot, 0);

    if (type != MELON_TYPE_NONE && arg->type != type)
    {
        if (!optional || arg->type != MELON_TYPE_NULL)
        {
            melM_fatal(
                vm, 
                "The %s should be a %s but was a '%s'",
                name,
                MELON_TYPES_NAMES[type],
                MELON_TYPES_NAMES[arg->type]
            );
            
            return NULL;
        }
    }

    return arg;
}

Value* melGetThisModule(VM* vm, const char* name)
{
    Value* arg = melGetLocalVM(vm, melGetTopCallFrameVM(vm), 0, 1);

    if (arg == NULL || arg->type != MELON_TYPE_OBJECT)
    {
        melM_fatal(
            vm, 
            "Can't get this for %s",
            name
        );

        return NULL;
    }

    return arg;
}