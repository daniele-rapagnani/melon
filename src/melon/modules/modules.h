#ifndef __melon__modules_h__
#define __melon__modules_h__

#include "melon/core/types.h"
#include "melon/core/vm.h"

#define melM_this(vm, name) \
    Value* name = melGetThisModule(vm, #name)

#define melM_arg(vm,name, type, slot) \
    Value* name = melGetArgModule(vm, #name, slot, type, 0)

#define melM_argOptional(vm,name, type, slot) \
    Value* name = melGetArgModule(vm, #name, slot, type, 1)

typedef TRet(*ModuleInitFunc)(VM*);

typedef struct Module
{
    const char* name;
    ModuleInitFunc initFunc;
} Module;

typedef struct ModuleFunction
{
    const char* name;
    TUint16 args;
    TUint16 localSlots;
    NativeFunctionPtr func;
    TBool method;
    TBool rest;
    MelonObjectSymbols symbol;
} ModuleFunction;

TRet melNewModule(VM* vm, const ModuleFunction* functions);
TRet melNewModuleFromSource(VM* vm, GCItem* path, GCItem* src, GCItem* errorCallback);
TRet melRegisterModule(VM* vm, const char* name, ModuleInitFunc initFunc);
TRet melRegisterModules(VM* vm, const Module* modules);

TRet melCreateKeyModule(VM* vm, const char* key, Value* value);
Value* melGetTypeFromModule(VM* vm, Value* moduleName, Value* keyName, TType type);
Value* melGetClosureFromModule(VM* vm, Value* moduleName, Value* funcName);

TRet melAddValueToObjModule(VM* vm, GCItem* obj, GCItem* name, Value* v);
TRet melAddItemToObjModule(VM* vm, GCItem* obj, GCItem* name, GCItem* i);
TRet melAddFunctionToObjModule(VM* vm, GCItem* obj, GCItem* name, GCItem* fn);
TRet melAddValueToGlobalModule(VM* vm, GCItem* name, Value* v);
TRet melAddFunctionToGlobalModule(VM* vm, GCItem* name, GCItem* fn);
TRet melAddItemToGlobalModule(VM* vm, GCItem* name, GCItem* item);
Value* melGetValueFromGlobalModule(VM* vm, Value* key);

GCItem* melCompileSourceModule(VM* vm, GCItem* path, GCItem* src, GCItem* errorCallback);
GCItem* melCompileBytecodeModule(VM* vm, GCItem* path, GCItem* code, GCItem* errorCallback);

Value* melGetArgModule(VM* vm, const char* name, TSize slot, TType type, TBool optional);
Value* melGetThisModule(VM* vm, const char* name);

#endif // __melon__modules_h__