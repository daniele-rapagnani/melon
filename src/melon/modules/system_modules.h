#ifndef __melon__system_modules_h__
#define __melon__system_modules_h__

#include "melon/modules/modules.h"

#include "melon/modules/io/io_module.h"
#include "melon/modules/import/import_module.h"
#include "melon/modules/string/string_module.h"
#include "melon/modules/fs/fs_module.h"
#include "melon/modules/path/path_module.h"
#include "melon/modules/math/math_module.h"
#include "melon/modules/json/json_module.h"
#include "melon/modules/object/object_module.h"
#include "melon/modules/types/types_module.h"
#include "melon/modules/compiler/compiler_module.h"
#include "melon/modules/debug/debug_module.h"
#include "melon/modules/cli/cli_module.h"
#include "melon/modules/number/number_module.h"
#include "melon/modules/integer/integer_module.h"
#include "melon/modules/array/array_module.h"
#include "melon/modules/function/function_module.h"
#include "melon/modules/gc/gc_module.h"
#include "melon/modules/random/random_module.h"
#include "melon/modules/system/system_module.h"

#include <stdlib.h>

Module MEL_SYSTEM_MODULES[] = {
    { "io", melIoModuleInit },
    { "fs", melFsModuleInit },
    { "path", melPathModuleInit },
    { "import", melImportModuleInit },
    { "string", melStringModuleInit },
    { "object", melObjectModuleInit },
    { "json", melJsonModuleInit },
    { "math", melMathModuleInit },
    { "types", melTypesModuleInit },
    { "compiler", melCompilerModuleInit },
    { "debug", melDebugModuleInit },
    { "cli", melCliModuleInit },
    { "number", melNumberModuleInit },
    { "integer", melIntegerModuleInit },
    { "array", melArrayModuleInit },
    { "function", melFunctionModuleInit },
    { "gc", melGcModuleInit },
    { "random", melRandomModuleInit },
    { "system", melSystemModuleInit },
    { NULL, NULL }
};

#endif // __melon__system_modules_h__