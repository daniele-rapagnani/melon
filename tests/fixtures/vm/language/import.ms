for (let p in import.paths) {
    io.print(p);
}

let obj = import("modules/test_module");
obj.doStuff();

let obj2 = import("modules/test_module");
obj2.doStuff();

import("modules/side_effects");
import("modules/side_effects");

import.cache = {};

import("modules/side_effects");

let moduleTest = import("modules/module_obj");
moduleTest->printInfo();

io.print("Number of cached modules:");
io.print(#import.cache);

import.paths = [
    "##/modules/" .. string.toString(system.getArchBits()) .. "/??.mbc",
    "##/??/index.ms",
    "##/../??/index.ms"
];

import.cache = {};

moduleTest = import("module_obj");
moduleTest->printInfo();

let oldImport = import[object.symbols.callOperator];

import[object.symbols.callOperator] = |path| => {
    io.print("Importing: " .. path);
    let importPath = null;
    let cs = debug.getCallstack();
    
    if (#cs >= 2) {
        importPath = function.getFile(cs[#cs - 2]);
    }

    return oldImport(path, importPath);
};

import.paths = [ "##/??.ms" ];

import("modules/test_module");
import("modules/test_module");

let cachedObj = import("modules/cache");
cachedObj.cached = "This value was cached";

cachedObj = import("modules/././cache");
io.print(cachedObj.cached);

import("modules/cached_user");
io.print(cachedObj.cached2);