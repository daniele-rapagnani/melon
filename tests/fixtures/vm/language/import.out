##/??
##/??.ms
##/??.mbc
##/??/index.ms
##/??/index.mbc
$$/mel_seeds/??.ms
$$/mel_seeds/??.mbc
$$/mel_seeds/??/index.ms
$$/mel_seeds/??/index.mbc
$$/mel_seeds/??
??
Should only print once!
Cool Module!
Cool Module!
This is a side-effect
This is a side-effect
module_obj.ms
Submodule found
Number of cached modules:
3
module_obj.mbc
Submodule found
Importing: modules/test_module
Should only print once!
Importing: modules/test_module
Importing: modules/cache
Importing: modules/././cache
This value was cached
Importing: modules/cached_user
Importing: ./cache
This value was cached
Another cached value
